File: jfx-webAuthn/src/main/scala-3/jfx/webauthn/WebAuthn.scala
package jfx.webauthn

import scala.concurrent.{ExecutionContext, Future}
import scala.scalajs.js
import scala.scalajs.js.JSConverters.*

object WebAuthn {

  private given ExecutionContext = ExecutionContext.global

  /** Whether the browser exposes both PublicKeyCredential and navigator.credentials. */
  def isSupported: Boolean =
    publicKeyCredentialApi.nonEmpty && credentialsContainer.nonEmpty

  /** Whether WebAuthn can run in the current secure browser context. */
  def isAvailable: Boolean =
    isSupported && {
      val secure = globalObject.selectDynamic("isSecureContext")
      js.isUndefined(secure) || secure.asInstanceOf[Boolean]
    }

  def isUserVerifyingPlatformAuthenticatorAvailable(): Future[Boolean] =
    isUserVerifyingPlatformAuthenticatorAvailablePromise().toFuture

  def isUserVerifyingPlatformAuthenticatorAvailablePromise(): js.Promise[Boolean] =
    booleanCapability("isUserVerifyingPlatformAuthenticatorAvailable")

  def isConditionalMediationAvailable(): Future[Boolean] =
    isConditionalMediationAvailablePromise().toFuture

  def isConditionalMediationAvailablePromise(): js.Promise[Boolean] =
    booleanCapability("isConditionalMediationAvailable")

  /** Returns all capabilities disclosed by the browser. Missing capabilities are not inferred. */
  def clientCapabilities(): Future[Map[String, Boolean]] =
    clientCapabilitiesPromise().toFuture.map(_.toMap)

  def clientCapabilitiesPromise(): js.Promise[js.Dictionary[Boolean]] =
    clientCapabilitiesValue().`then`[js.Dictionary[Boolean]]((value: js.Any) => {
      val source = value.asInstanceOf[js.Dynamic]
      val result = js.Dictionary.empty[Boolean]
      js.Object.keys(value.asInstanceOf[js.Object]).foreach { key =>
        val capability = source.selectDynamic(key)
        if (js.typeOf(capability) == "boolean")
          result.update(key, capability.asInstanceOf[Boolean])
      }
      result
    })

  def register(options: js.Dynamic): Future[RegistrationCredential] =
    registerPromise(options).toFuture

  def register(
      options: PublicKeyCredentialCreationOptions,
      settings: CredentialCreationSettings = CredentialCreationSettings()
  ): Future[RegistrationCredential] =
    registerPromise(options, settings).toFuture

  def registerJson(
      options: js.Dynamic,
      settings: CredentialCreationSettings = CredentialCreationSettings()
  ): Future[RegistrationCredential] =
    registerJsonPromise(options, settings).toFuture

  def registerJson(
      json: String,
      settings: CredentialCreationSettings
  ): Future[RegistrationCredential] =
    registerJsonPromise(json, settings).toFuture

  def registerJson(json: String): Future[RegistrationCredential] =
    registerJson(json, CredentialCreationSettings())

  def authenticate(options: js.Dynamic): Future[AuthenticationCredential] =
    authenticatePromise(options).toFuture

  def authenticate(
      options: PublicKeyCredentialRequestOptions,
      settings: CredentialRequestSettings = CredentialRequestSettings()
  ): Future[AuthenticationCredential] =
    authenticatePromise(options, settings).toFuture

  def authenticateJson(
      options: js.Dynamic,
      settings: CredentialRequestSettings = CredentialRequestSettings()
  ): Future[AuthenticationCredential] =
    authenticateJsonPromise(options, settings).toFuture

  def authenticateJson(
      json: String,
      settings: CredentialRequestSettings
  ): Future[AuthenticationCredential] =
    authenticateJsonPromise(json, settings).toFuture

  def authenticateJson(json: String): Future[AuthenticationCredential] =
    authenticateJson(json, CredentialRequestSettings())

  def registerPromise(options: js.Dynamic): js.Promise[RegistrationCredential] =
    registerJsonPromise(options)

  def registerJsonPromise(options: js.Dynamic): js.Promise[RegistrationCredential] =
    registerJsonPromise(options, CredentialCreationSettings())

  def registerJsonPromise(
      options: js.Dynamic,
      settings: CredentialCreationSettings
  ): js.Promise[RegistrationCredential] =
    attemptPromise(registerPromise(WebAuthnCodecs.creationOptionsFromJson(options), settings))

  def registerJsonPromise(json: String): js.Promise[RegistrationCredential] =
    registerJsonPromise(json, CredentialCreationSettings())

  def registerJsonPromise(
      json: String,
      settings: CredentialCreationSettings
  ): js.Promise[RegistrationCredential] =
    attemptPromise(registerPromise(WebAuthnCodecs.creationOptionsFromJson(json), settings))

  def registerPromise(
      options: PublicKeyCredentialCreationOptions,
      settings: CredentialCreationSettings = CredentialCreationSettings()
  ): js.Promise[RegistrationCredential] =
    credentialOperation("create", CredentialCreationOptions.withSettings(options, settings))
      .`then`[RegistrationCredential]((credential: PublicKeyCredential) =>
        WebAuthnCodecs.registrationCredential(credential)
      )

  def authenticatePromise(options: js.Dynamic): js.Promise[AuthenticationCredential] =
    authenticateJsonPromise(options)

  def authenticateJsonPromise(options: js.Dynamic): js.Promise[AuthenticationCredential] =
    authenticateJsonPromise(options, CredentialRequestSettings())

  def authenticateJsonPromise(
      options: js.Dynamic,
      settings: CredentialRequestSettings
  ): js.Promise[AuthenticationCredential] =
    attemptPromise(authenticatePromise(WebAuthnCodecs.requestOptionsFromJson(options), settings))

  def authenticateJsonPromise(json: String): js.Promise[AuthenticationCredential] =
    authenticateJsonPromise(json, CredentialRequestSettings())

  def authenticateJsonPromise(
      json: String,
      settings: CredentialRequestSettings
  ): js.Promise[AuthenticationCredential] =
    attemptPromise(authenticatePromise(WebAuthnCodecs.requestOptionsFromJson(json), settings))

  def authenticatePromise(
      options: PublicKeyCredentialRequestOptions,
      settings: CredentialRequestSettings = CredentialRequestSettings()
  ): js.Promise[AuthenticationCredential] =
    credentialOperation("get", CredentialRequestOptions.withSettings(options, settings))
      .`then`[AuthenticationCredential]((credential: PublicKeyCredential) =>
        WebAuthnCodecs.authenticationCredential(credential)
      )

  def signalUnknownCredential(options: UnknownCredentialOptions): Future[Unit] =
    signalUnknownCredentialPromise(options).toFuture.map(_ => ())

  def signalUnknownCredentialPromise(options: UnknownCredentialOptions): js.Promise[Unit] =
    callStaticPromise("signalUnknownCredential", js.Array(options.toJsObject)).`then`[Unit](_ => ())

  def signalAllAcceptedCredentials(options: AllAcceptedCredentialsOptions): Future[Unit] =
    signalAllAcceptedCredentialsPromise(options).toFuture.map(_ => ())

  def signalAllAcceptedCredentialsPromise(
      options: AllAcceptedCredentialsOptions
  ): js.Promise[Unit] =
    callStaticPromise("signalAllAcceptedCredentials", js.Array(options.toJsObject)).`then`[Unit](
      _ => ()
    )

  def signalCurrentUserDetails(options: CurrentUserDetailsOptions): Future[Unit] =
    signalCurrentUserDetailsPromise(options).toFuture.map(_ => ())

  def signalCurrentUserDetailsPromise(options: CurrentUserDetailsOptions): js.Promise[Unit] =
    callStaticPromise("signalCurrentUserDetails", js.Array(options.toJsObject)).`then`[Unit](_ =>
      ()
    )

  private def credentialOperation(
      methodName: String,
      options: js.Object
  ): js.Promise[PublicKeyCredential] =
    credentialsContainer match {
      case None            => unsupportedPromise("WebAuthn is not supported in this environment")
      case Some(container) =>
        val method = container.selectDynamic(methodName)
        if (js.typeOf(method) != "function")
          unsupportedPromise(s"navigator.credentials.$methodName is not supported")
        else
          method
            .call(container, options)
            .asInstanceOf[js.Promise[PublicKeyCredential | Null]]
            .`then`[PublicKeyCredential] { credential =>
              Option(credential).getOrElse {
                throw new IllegalStateException(
                  s"navigator.credentials.$methodName returned no credential"
                )
              }
            }
    }

  private def booleanCapability(name: String): js.Promise[Boolean] =
    publicKeyCredentialApi match {
      case None      => js.Promise.resolve(false)
      case Some(api) =>
        val method = api.selectDynamic(name)
        if (js.typeOf(method) != "function") js.Promise.resolve(false)
        else method.call(api).asInstanceOf[js.Promise[Boolean]]
    }

  private def clientCapabilitiesValue(): js.Promise[js.Any] =
    publicKeyCredentialApi match {
      case None      => js.Promise.resolve(js.Dynamic.literal())
      case Some(api) =>
        val method = api.selectDynamic("getClientCapabilities")
        if (js.typeOf(method) != "function") js.Promise.resolve(js.Dynamic.literal())
        else method.call(api).asInstanceOf[js.Promise[js.Any]]
    }

  private def callStaticPromise(
      name: String,
      arguments: js.Array[js.Any]
  ): js.Promise[js.Any] =
    publicKeyCredentialApi match {
      case None      => unsupportedPromise("WebAuthn is not supported in this environment")
      case Some(api) =>
        val method = api.selectDynamic(name)
        if (js.typeOf(method) != "function")
          unsupportedPromise(s"PublicKeyCredential.$name is not supported")
        else method.apply(api, arguments).asInstanceOf[js.Promise[js.Any]]
    }

  private def credentialsContainer: Option[js.Dynamic] = {
    val navigator = globalObject.selectDynamic("navigator")
    if (navigator == null || js.isUndefined(navigator)) None
    else {
      val credentials = navigator.selectDynamic("credentials")
      if (credentials == null || js.isUndefined(credentials)) None
      else Some(credentials.asInstanceOf[js.Dynamic])
    }
  }

  private def publicKeyCredentialApi: Option[js.Dynamic] = {
    val api = globalObject.selectDynamic("PublicKeyCredential")
    if (api == null || js.isUndefined(api)) None else Some(api.asInstanceOf[js.Dynamic])
  }

  private def globalObject: js.Dynamic =
    js.Dynamic.global.selectDynamic("globalThis").asInstanceOf[js.Dynamic]

  private def unsupportedPromise[A](message: String): js.Promise[A] =
    js.Promise.reject(new UnsupportedOperationException(message))

  private def attemptPromise[A](operation: => js.Promise[A]): js.Promise[A] =
    try operation
    catch {
      case error: Throwable => js.Promise.reject(error)
    }
}

File: npm/scalajs-ui-webauthn/src/webauthn.ts
import {
  authenticationCredential,
  creationOptionsFromJson,
  requestOptionsFromJson,
  registrationCredential,
  type AuthenticationCredential,
  type PublicKeyCredentialCreationOptionsJSON,
  type PublicKeyCredentialRequestOptionsJSON,
  type RegistrationCredential,
} from "./codecs.js";
import {
  credentialCreationOptions,
  credentialRequestOptions,
  type AllAcceptedCredentialsOptions,
  type CredentialCreationSettings,
  type CredentialRequestSettings,
  type CurrentUserDetailsOptions,
  type PublicKeyCredentialCreationOptions,
  type PublicKeyCredentialRequestOptions,
  type UnknownCredentialOptions,
} from "./facades.js";

/** Whether the browser exposes both PublicKeyCredential and navigator.credentials. */
export function isSupported(): boolean {
  return publicKeyCredentialApi() !== undefined && credentialsContainer() !== undefined;
}

/** Whether WebAuthn can run in the current secure browser context. */
export function isAvailable(): boolean {
  const secure = globalThis.isSecureContext;
  return isSupported() && (secure === undefined || secure);
}

export async function isUserVerifyingPlatformAuthenticatorAvailable(): Promise<boolean> {
  return isUserVerifyingPlatformAuthenticatorAvailablePromise();
}

export function isUserVerifyingPlatformAuthenticatorAvailablePromise(): Promise<boolean> {
  return booleanCapability("isUserVerifyingPlatformAuthenticatorAvailable");
}

export async function isConditionalMediationAvailable(): Promise<boolean> {
  return isConditionalMediationAvailablePromise();
}

export function isConditionalMediationAvailablePromise(): Promise<boolean> {
  return booleanCapability("isConditionalMediationAvailable");
}

/** Returns all capabilities disclosed by the browser. Missing capabilities are not inferred. */
export async function clientCapabilities(): Promise<Record<string, boolean>> {
  return clientCapabilitiesPromise();
}

export async function clientCapabilitiesPromise(): Promise<Record<string, boolean>> {
  const value = await clientCapabilitiesValue();
  const result: Record<string, boolean> = {};
  if (value && typeof value === "object") {
    for (const [key, capability] of Object.entries(value)) {
      if (typeof capability === "boolean") result[key] = capability;
    }
  }
  return result;
}

export function register(
  options: PublicKeyCredentialCreationOptions,
  settings: CredentialCreationSettings = {},
): Promise<RegistrationCredential> {
  return registerPromise(options, settings);
}

export function registerJson(
  options: PublicKeyCredentialCreationOptionsJSON | string,
  settings: CredentialCreationSettings = {},
): Promise<RegistrationCredential> {
  return registerJsonPromise(options, settings);
}

export function registerPromise(
  options: PublicKeyCredentialCreationOptions,
  settings: CredentialCreationSettings = {},
): Promise<RegistrationCredential> {
  return credentialOperation(
    "create",
    credentialCreationOptions(options, settings),
  ).then((credential) => registrationCredential(credential));
}

export function registerJsonPromise(
  options: PublicKeyCredentialCreationOptionsJSON | string,
  settings: CredentialCreationSettings = {},
): Promise<RegistrationCredential> {
  return attemptPromise(() => registerPromise(parseCreationOptions(options), settings));
}

export function authenticate(
  options: PublicKeyCredentialRequestOptions,
  settings: CredentialRequestSettings = {},
): Promise<AuthenticationCredential> {
  return authenticatePromise(options, settings);
}

export function authenticateJson(
  options: PublicKeyCredentialRequestOptionsJSON | string,
  settings: CredentialRequestSettings = {},
): Promise<AuthenticationCredential> {
  return authenticateJsonPromise(options, settings);
}

export function authenticatePromise(
  options: PublicKeyCredentialRequestOptions,
  settings: CredentialRequestSettings = {},
): Promise<AuthenticationCredential> {
  return credentialOperation(
    "get",
    credentialRequestOptions(options, settings),
  ).then((credential) => authenticationCredential(credential));
}

export function authenticateJsonPromise(
  options: PublicKeyCredentialRequestOptionsJSON | string,
  settings: CredentialRequestSettings = {},
): Promise<AuthenticationCredential> {
  return attemptPromise(() => authenticatePromise(parseRequestOptions(options), settings));
}

export async function signalUnknownCredential(options: UnknownCredentialOptions): Promise<void> {
  await signalUnknownCredentialPromise(options);
}

export function signalUnknownCredentialPromise(options: UnknownCredentialOptions): Promise<void> {
  return callStaticPromise("signalUnknownCredential", options);
}

export async function signalAllAcceptedCredentials(
  options: AllAcceptedCredentialsOptions,
): Promise<void> {
  await signalAllAcceptedCredentialsPromise(options);
}

export function signalAllAcceptedCredentialsPromise(
  options: AllAcceptedCredentialsOptions,
): Promise<void> {
  return callStaticPromise("signalAllAcceptedCredentials", options);
}

export async function signalCurrentUserDetails(options: CurrentUserDetailsOptions): Promise<void> {
  await signalCurrentUserDetailsPromise(options);
}

export function signalCurrentUserDetailsPromise(
  options: CurrentUserDetailsOptions,
): Promise<void> {
  return callStaticPromise("signalCurrentUserDetails", options);
}

function parseCreationOptions(
  value: PublicKeyCredentialCreationOptionsJSON | string,
): PublicKeyCredentialCreationOptions {
  return creationOptionsFromJson(typeof value === "string" ? parseJsonObject(value) : value);
}

function parseRequestOptions(
  value: PublicKeyCredentialRequestOptionsJSON | string,
): PublicKeyCredentialRequestOptions {
  return requestOptionsFromJson(typeof value === "string" ? parseJsonObject(value) : value);
}

function credentialOperation(
  methodName: "create" | "get",
  options: object,
): Promise<PublicKeyCredential> {
  const container = credentialsContainer();
  if (!container) return unsupportedPromise("WebAuthn is not supported in this environment");

  const method = container[methodName];
  if (typeof method !== "function") {
    return unsupportedPromise(`navigator.credentials.${methodName} is not supported`);
  }

  return Promise.resolve(
    method.call(container, options as unknown as globalThis.CredentialCreationOptions),
  ).then((credential) => {
    if (!credential) {
      throw new Error(`navigator.credentials.${methodName} returned no credential`);
    }
    return credential as PublicKeyCredential;
  });
}

function booleanCapability(name: string): Promise<boolean> {
  const api = publicKeyCredentialApi();
  const method = api?.[name];
  if (typeof method !== "function") return Promise.resolve(false);
  return Promise.resolve(method.call(api)).then((value) => value === true);
}

function clientCapabilitiesValue(): Promise<unknown> {
  const api = publicKeyCredentialApi();
  const method = api?.["getClientCapabilities"];
  if (typeof method !== "function") return Promise.resolve({});
  return Promise.resolve(method.call(api));
}

function callStaticPromise(name: string, options: unknown): Promise<void> {
  const api = publicKeyCredentialApi();
  if (!api) return unsupportedPromise("WebAuthn is not supported in this environment");

  const method = api[name];
  if (typeof method !== "function") {
    return unsupportedPromise(`PublicKeyCredential.${name} is not supported`);
  }
  return Promise.resolve(method.call(api, options)).then(() => undefined);
}

function credentialsContainer(): Record<string, unknown> | undefined {
  const navigatorValue = globalThis.navigator;
  const credentials = navigatorValue?.credentials;
  return credentials && typeof credentials === "object"
    ? credentials as unknown as Record<string, unknown>
    : undefined;
}

function publicKeyCredentialApi(): Record<string, unknown> | undefined {
  const api = globalThis.PublicKeyCredential;
  return api && typeof api === "function"
    ? api as unknown as Record<string, unknown>
    : undefined;
}

function parseJsonObject(json: string): Record<string, unknown> {
  try {
    const value: unknown = JSON.parse(json);
    if (value === null || typeof value !== "object" || Array.isArray(value)) {
      throw new TypeError("Invalid WebAuthn JSON root: object expected");
    }
    return value as Record<string, unknown>;
  } catch (error) {
    if (error instanceof TypeError && error.message.includes("object expected")) throw error;
    throw new TypeError("Invalid JSON", { cause: error });
  }
}

function attemptPromise<T>(operation: () => Promise<T>): Promise<T> {
  try {
    return operation();
  } catch (error) {
    return Promise.reject(error);
  }
}

function unsupportedPromise<T>(message: string): Promise<T> {
  return Promise.reject(new Error(message));
}

export const WebAuthn = {
  isSupported,
  isAvailable,
  isUserVerifyingPlatformAuthenticatorAvailable,
  isUserVerifyingPlatformAuthenticatorAvailablePromise,
  isConditionalMediationAvailable,
  isConditionalMediationAvailablePromise,
  clientCapabilities,
  clientCapabilitiesPromise,
  register,
  registerJson,
  registerPromise,
  registerJsonPromise,
  authenticate,
  authenticateJson,
  authenticatePromise,
  authenticateJsonPromise,
  signalUnknownCredential,
  signalUnknownCredentialPromise,
  signalAllAcceptedCredentials,
  signalAllAcceptedCredentialsPromise,
  signalCurrentUserDetails,
  signalCurrentUserDetailsPromise,
} as const;
